contracts/sol4/bridgeReserves/uniswapCurve/KyberUniswapCurveReserve.sol

```diff
@@ -0,0 +1,435 @@
+            srcToken,
+            destToken,
+            srcAmount,
+            conversionRate
+        );
+
+        ERC20 bridgeToken;
+        bool useCurve; // whether using Curve is better
+        uint256 destAmount;
+        uint256 uniswapDestAmount;
+        (bridgeToken, useCurve, destAmount, uniswapDestAmount) =
+            getTradeInformation(srcToken, destToken, srcAmount);
+
+        require(expectedDestAmount <= destAmount);
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            destAmount = doTradeEthToToken(
+                destToken,
+                bridgeToken,
+                useCurve,
+                srcAmount,
+                uniswapDestAmount
+            );
+            require(destAmount >= expectedDestAmount);
+            destToken.transfer(destAddress, expectedDestAmount);
+        } else {
+            // collect src amount
+            srcToken.transferFrom(msg.sender, address(this), srcAmount);
+            destAmount = doTradeTokenToEth(
+                srcToken,
+                bridgeToken,
+                useCurve,
+                srcAmount,
+                uniswapDestAmount
+            );
+            require(destAmount >= expectedDestAmount);
+            destAddress.transfer(expectedDestAmount);
+        }
+
+        TradeExecute(
+            msg.sender,
+            srcToken,
+            srcAmount,
+            destToken,
+            expectedDestAmount,
+            destAddress
+        );
+        return true;
+    }
+
+    /**
+     *   @dev called by kybernetwork to get settlement rate
+     */
+    function getConversionRate(
+        ERC20 src,
+        ERC20 dest,
+        uint256 srcQty,
+        uint256 /* blockNumber */
+    ) public view returns (uint256) {
+        if (!isValidTokens(src, dest)) return 0;
+        if (!tradeEnabled) return 0;
+        if (srcQty == 0) return 0;
+
+        uint256 destAmount;
+        (, , destAmount, ) =
+            getTradeInformation(src, dest, srcQty);
+        if (destAmount == 0) return 0;
+        return calcRateFromQty(
+            srcQty,
+            destAmount,
+            getDecimals(src),
+            getDecimals(dest)
+        );
+    }
+
+    function getTradeInformation(ERC20 src, ERC20 dest, uint256 srcQty)
+        public view
+        returns(
+            ERC20 bridgeToken,
+            bool useCurve,
+            uint256 destAmount,
+            uint256 uniswapDestAmount
+        )
+    {
+        address[] memory tokens;
+        ERC20 token;
+        uint256 i;
+        uint256 destQty;
+        uint256 destUniswapQty;
+
+        if (src == ETH_TOKEN_ADDRESS) {
+            // check eth -> token in Uniwap, token -> dest in Curve
+            // first, not use Curve, get amount eth-> dest in Uniswap
+            uniswapDestAmount = getUniswapDestAmount(dest, srcQty, true);
+            destAmount = uniswapDestAmount;
+            useCurve = false;
+            tokens = bridgeTokens[dest];
```

Comment:
its unclear how important it is to have more than 1 bridge tokens.
since it means much more gas for the trade.
how big is the diff between both bridge tokens?
could it change a lot?

combined rate query cost I see here is 125K gas
https://dashboard.tenderly.co/tx/main/0x0dd16e9840f769a61672754f85c7ce7d28bd2c9097f2aff36e9bf91f79d653db/gas-usage

Reply:
We don't know which bridge token is the best for the trade, we only add tokens with high liquidy in eth-token pair on Uniswap, for example: eth -> usdt, we can trade eth -> usdc (Uniswap) + usdc -> usdt (Curve) or eth -> dai (Uniswap) + dai -> usdt (Curve). Sometimes using dai is better, sometimes using usdc is better.

Reply:
Where does it show that the combined rate query costs 125k gas?

Reply:
if opening the shared link
and then pressing on calcRatesAndAmounts

the highest getRate call is from this reserve